### PR TITLE
weapondef option to update query pieces faster

### PR DESCRIPTION
**Pull request by KyleAnthonyShepherd**
Currently, weapon pieces are only updated
1. Immediately before the weapon projectile is created
2. On SlowUpdate frames

This causes significant friendly fire problems for rapid-fire (<0.5 sec) multi-barrel units. 
1. unit wants to fire from barrel 1
2. friendly fire checks are done from barrel 1
3. FireWeaponX unitscript function is called
4. weapon pieces updated in Weapon.cpp
5. Projectile fired from barrel 1
6. FireWeaponX unitscript function switches to barrel 2 a couple of frames later
7. Engine still does friendly fire checks from barrel 1, because it does not ask the unitscript if the weapon pieces have been updated
8. FireWeaponX unitscript function is called
9. weapon pieces updated [barrel 1 -> barrel 2]
10. Projectile fired from barrel 2, potentially colliding with friendly units.

This PR addresses this problem by updating weapon pieces in CWeapon::Update() before friendly fire checks are performed. 
As this may change existing unit behavior, and incur a performance cost due to updating weapon pieces every frame, a bool in the weapons subtable of the UnitDef (along the lines of the recently added fastAutoRetargeting bool) is added to enable this behavior on a unit-by-unit basis.

### CODE REVIEW

**Comment by WatchTheFort** [comment]
Does this need a separate parameter? Any reason not to tie this to the `fastAutoRetargeting` bool?

**Comment by sprunk** [comment]
> Any reason not to tie this to the fastAutoRetargeting bool?

The two are orthogonal and both carry extra perf costs. `fastAutoRetargeting` is for anything that kills targets quickly after acquisition (including single-barrel weapons) and `fastQueryPointUpdate` is for anything that has many barrels and switches between them rapidly (including weapons that aren't expected to kill their targets quickly after acquisiton, due to low damage or being non-hitscan).

<hr/>

Does the PR correctly address weapons that fire from multiple barrels on the same frame (via `projectiles = N` in the weapon def and adjusting the barrel between them in the script's `Shot` function)? 

I think I would like to request an explicit `Spring.UpdateWeaponQueryPoint(unitID, weaponNum)` callout since a gamedev knows best when there's a need to update pieces. Consider a setup where a unit has a powerful, long reload gun with 3 barrels that all shoot simultaneously (think a battleship).

Even with this feature there's still a lot to be desired:
 * unit shoots barrel 1, switches to barrel 2 in script.
 * unit shoots barrel 2, **but engine still checks friendly fire from barrel 1 because there was no frame inbetween**. Switches to barrel 3 in script.
 * unit shoots barrel 3, **but engine still checks friendly fire from barrel 2 because there was no frame inbetween**. Switches to barrel 1 in script.

Or, consider a similar weapon but there's a 1 frame delay between the barrels shooting to leave room for the fast update:
 * unit shoots barrel 1, switches to barrel 2 in script.
 * waits 1 frame, during which the engine updates friendly fire query point to barrel 2 correctly.
 * unit shoots barrel 2, switches to barrel 3 in script.
 * waits 1 frame, during which the engine updates friendly fire query point to barrel 3 correctly.
 * unit shoots barrel 3, switches to barrel 1 in script.
 * waits 300 frames for a reload **and does a useless weapon piece update in every single one of those frames**.

With an explicit callout it works nicely:
 * unit shoots barrel 1, switches to barrel 2 in script **and lets the engine know about it via the callout**.
 * unit shoots barrel 2, switches to barrel 3 in script **and lets the engine know about it via the callout**.
 * unit shoots barrel 3, switches to barrel 1 in script **and lets the engine know about it via the callout**.
 * waits 300 frames for a reload without extra overhead.

**Comment by KyleAnthonyShepherd** [comment]
https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Sim/Weapons/Weapon.cpp#L516

Currently, only ever the first projectile in a burst OR salvo goes through friendly fire checks. 
First in salvo because a successful `UpdateFire()` sets `salvoLeft>1`, and so allows `UpdateSalvo()` to run multiple times.
First in burst because of the for loop in L516 that only calls the unit Shot script, updates weapon pieces, and creates the projectile. 

`fastQueryPointUpdate` just makes sure that this first in burst or salvo projectile does the correct friendly fire check if the reload time is less than 0.5 seconds.

A few ambiguities would have to be addressed if the additional shots in a salvo or burst are checked for friendly fire.
Such as if shot 3 of 5 of a salvo is blocked by friendly fire, does the salvo stop (never firing shots 3, 4, 5), does it wait (delay firing shots 3, 4, 5 until a clear trajectory), or does it skip (never firing shot 3, then checking friendly fire for shots 4 and 5)

**Comment by sprunk** [comment]
> Currently, only ever the first projectile in a burst OR salvo goes through friendly fire checks.

Right, disregard most of the above then. A manual callout would still be good though because you can still get long periods where the weapon doesn't fire and it's a waste of perf to keep updating every frame, but the current weapon tag is also fine.

**Comment by lhog** [comment]
Am I getting it right that with `fastQueryPointUpdate = true` in place the `script->QueryWeapon()` is going to be called every frame regardless of reload state?

**Comment by KyleAnthonyShepherd** [comment]
> Am I getting it right that with `fastQueryPointUpdate = true` in place the `script->QueryWeapon()` is going to be called every frame regardless of reload state?

That is correct. It needs to be called before `UpdateAim()` so the checks are using the correct point.
It may be worth it to put it after the HaveTarget check to reduce how often `script->QueryWeapon()` is called.

**Comment by KyleAnthonyShepherd** [comment]
Correction, I looked again, and positioning `UpdateWeaponPieces(false)` and `UpdateWeaponVectors()` after `CanFire`, and before `TryTarget` does work.

So the query point updating is only performed when the weapon is reloaded and could fire.

File: rts/Sim/Weapons/Weapon.cpp [hunk 1]
```diff
@@ -436,6 +433,10 @@ void CWeapon::UpdateFire()
 	if (!CanFire(false, false, false))
 		return;
 
+	if (fastQueryPointUpdate)
+		UpdateWeaponPieces(false);
+		UpdateWeaponVectors();
```
Comment (sprunk) [inline rts/Sim/Weapons/Weapon.cpp:438]: brackets? right now `UpdateWeaponVectors();` happen unconditionally

**Review by sprunk — CHANGES_REQUESTED** [review]

File: rts/Sim/Weapons/Weapon.cpp (hunk 1 above)
Comment (KyleAnthonyShepherd) [reply rts/Sim/Weapons/Weapon.cpp]: Added

**Review by sprunk — APPROVED** [review]

**Comment by lhog** [comment]
The implementation https://github.com/beyond-all-reason/spring/pull/901/commits/fe8ee2b8b3b6382c8aa4503300b5d7a1138feabe looks much better placed than before.
I thought I was going to propose to switch over the implementation from pull to push, like @sprunk suggested, but now it seems acceptable, since calls only occur during the "firing" stage.

**Pull request merged**